inspect_dynspec/gpr_smooth.py

```
@@ -0,0 +1,266 @@
+
+
+def gpr_smooth(
+    data: jnp.ndarray,  # shape (Nv, Nt)
+    mask: jnp.ndarray,  # shape (Nv, Nt), bool or {0,1}
+    weights: Optional[
+        jnp.ndarray
+    ] = None,  # shape (Nv, Nt), per-pixel noise-precision (1/σ²)
+    l_length_nu: float = 1.0,
+    l_length_t: float = 1.0,
+    sigma2: float = 1.0,
+    jitter: float = 1e-6,
+    cg_tol: float = 1e-6,
+    cg_maxiter: int = 200,
+    nof_weight_samples: int = 20,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """
+    2D Gaussian‐Process smoothing via an RBF kernel with optional heteroscedastic noise.
+
+    Args:
+      data     : float[H, W]  — observed brightness.
+      mask     : float[H, W]  — per-pixel mask (1 for valid pixels, 0 for invalid).
+      weights  : float[H, W]  — per-pixel noise-precision (1/sigma^2). If None, uniform weights are used.
+      l_length_nu : float     — RBF length-scale for frequency axis.
+      l_length_t : float      — RBF length-scale for time axis.
+      sigma2   : float        — RBF signal variance sigma^2.
+      jitter   : float        — small value added to diagonal for numerical stability.
+      cg_tol   : float        — CG solver tolerance.
+      cg_maxiter: int         — maximum number of CG iterations.
+      nof_weight_samples: int — number of latent samples to estimate posterior variance.
+
+    Returns:
+      float[H, W] — posterior mean (“smoothed”) image.
+      float[H, W] — posterior variance image.
+    """
+
+    # Prep:
+    Nv, Nt = data.shape
+    t_grid = jnp.linspace(0, 1, Nt, dtype=jnp.float64)
+    v_grid = jnp.linspace(0, 1, Nv, dtype=jnp.float64)
+    R = Mask(mask)
+
+    weights = jnp.ones_like(data) if weights is None else weights
+
+    data_flat = data.ravel()
+    prec_flat = weights.ravel()
+
+    y_obs = R.forward(data_flat)  # observed data
+    prec_obs = R.forward(prec_flat)  # observed precisions
+
+    # Σ = diag(1/prec_flat)  ⇒  σ_obs = 1/√prec_obs
+    sigma_obs = 1.0 / jnp.sqrt(jnp.where(prec_obs == 0, 1, prec_obs))
```

Ideally the mask operator should remove entries with zero weight so the `where` shouldn't be necessary. Might be better to raise an error here if any `prec_obs` are zero